### ISSUE TITLE
set_flags_in_fixed_headers: choice on the computation of the range end

### ISSUE THREAD

**Issue by etmseed** [obspy/obspy#3255]
### Avoid duplicates

- [X] I searched existing issues

### Bug Summary

My code generates mseed files and calls the function obspy.io.mseed.util.set_flags_in_fixed_headers.
The flags to be set are of the 'DURATION' kind and I could observe in the resulting files that some records had the flags set when they should not wrt to their data dates.

I looked into the code of this function and I was surprised by the comment at lines obspy/io/mseed/util.py:1161-1162 in obspy-1.3.1:
```
                # We assume here that a record with samples [0, 1, ..., n]
                # has a period [ date_0, date_n+1 [  AND NOT [ date_0, date_n ]
```

Actually, on my example, I have an event that starts between date(n) and date(n+1) (a 10s range, since my signal has got a very low frequency of 0.1Hz !). This way of computing the realendtime includes the start of the event and I end up with a 30-min long record flagged with the event. From a scientific point of view, it means that the whole record is not relevant because the measures were impacted by that event.
On the contrary, all of the record data were got before the event started and are thus relevant !

My understanding is that the line 1163 should have been:
`                realendtime = recstart + (nb_samples - 1) / samp_rate`
and the line 1510:
`                    if (event_start < recend) and (recstart <= event_end):`
(instead of event_start < recend)
But this may imply to make sure nb_samples > 1...

I know this is clearly a deliberate choice but after I was confronted with the situation above, I cannot think of any example where this would be relevant and I have thus the feeling this is a bug.

### ObsPy Version?

1.0.2 and 1.3.1 (and probably all intermediates)

### Operating System?

CentOS 6

### Python Version?

2.7

### Installation Method?

pip

**Comment by megies** [comment]
I do not know all the details of record level MSEED flags and how people use those, but this seems like an edge case where you could argue both ways. We had this discussion pop up from time to times in other places.. what should be shown as endtime of a Trace? Time of last sample? Or time of last sample plus delta? If you choose the former, and then think about a trace with only one sample, you end up with your information spanning no time at all, even though you have a measurement and have information..
Let's say you do a measurement on climate with one sample per year and you have two samples, is your information really representing only the span of one year?

> and the line 1510:
> ` if (event_start < recend) and (recstart <= event_end):`
> (instead of event_start < recend)

Isn't that what is currently in the code?

---

This is just guessing from the above.. If I understand right you have some workflow that is based on miniseed flags written on record level and want to process data that has no earthquakes in it and you are worried about losing data (record being ignored due to set flags) for when an origin time falls into the time span of the record (here: around the last sample)? Does it really matter much? If you make your record length very long, you and up with many samples in one record and the chances that you hit that time span of one delta with your "event" time becomes really small, so the likelihood of losing that record due to this matter is really low. On the other hand you could choose a shorter record length, which makes the likelihood of losing records to this matter higher, but then the amount of data you lose every time it happens becomes less.

Or maybe I totally don't understand the underlying problem you seem to have in your workflow?

**Comment by etmseed** [comment]
I understand your point about the general meaning to be given to the end of a trace. Yet we are here on a very specific functionality which is the flagging of a record depending on event dates. It is somehow tainting samples with some information and not tainting a period of time with that information...
Wrt to the line 1510, I forgot the <=, of course. I fixed it in the original comment.

I'll try to give you clues about my reasons for this new issue. We receive data from a sensor. Some of them should never be used for scientific analysis or they could lead to wrong conclusions. These data are those collected while some operations were performed on the instrument, such as calibration or motor moves that do not saturate the measurements. Of course you could argue that scientists would never mistake the signal resulting from a motor move with an earthquake, but our specifications state that those data shall be marked/flagged in the produced mseed files.
The dates of such operations are retrieved independently from the sensor data and are processed as DURATION events by our software. In order to have the best possible accuracy, the software splits up the traces on event dates, creating records with only the samples that are actually impacted by the event/operation. The record right before the "event record" has usually got date(n+1) = date(0) of the event record, and the record right after has date(0) = date(n+1) of the event record.

The way realendtime is computed in that function leads to flag the record before the event record too.
I am quite new to mseed generation, but with what I read about those flags, I cannot think of any good reason to do it this way.
Again, I understand your point about the definition of the end of a trace. Yet flagging a record sounds to me like flagging samples and not flagging a time period, which makes that computation disconnected from the trace end concept.
But as I said, I am quite new to this and may be totally wrong on my understanding.

**Comment by megies** [comment]
> the software splits up the traces on event dates, creating records with only the samples that are actually impacted by the event/operation.

Can't you just tell your software to make that split one more sample to the left? That's basically what you're telling us to do, isn't it.. :wink:

**Comment by megies** [comment]
I mean, I can see your point to some extent.. and maybe it's a "whatever" to change it. Maybe somebody working more on MiniSEED record level flags could share their views on this? @Jollyfant maybe? Or @chad-earthscope?

**Comment by etmseed** [comment]
> > the software splits up the traces on event dates, creating records with only the samples that are actually impacted by the event/operation.
> 
> Can't you just tell your software to make that split one more sample to the left? That's basically what you're telling us to do, isn't it.. 😉

Yes I can, but...
1) I have the feeling it is a bit different from what I expected from obspy because it would lead to flagging one sample that is outside the event time window.
2) Different signals (temperature/pressure/seismic data) are compared altogether by the scientists. Those signals have different periods: seismic data may have a period of 1/100s while temperature may have a 10-second period. It is not an issue to flag one more sample on high-frequency signals but on low-frequency ones, it may lead to some confusions in comparison to those with short periods. It is expected that higher frequency signals would be flagged sooner than lower frequency ones, if you understand what I mean.

**Comment by chad-earthscope** [comment]
I do not believe I'll be able to add much to this discussion, and I'm not sure I understand all the details, but here's some thoughts.

The concept of bit flags describing duration type events in miniSEED is imperfect.  There is no choice but to assume the flags apply to all the samples in the record, whether that's true or not.  Short of manipulating the number of samples in a record, the concepts of duration and applicable samples will not match.

For me, sample values do not have duration; they are measurements at points in time, even though representing a continuous waveform.  Of course we need to represent their combined coverage for many practical reasons.  It is common to treat a sample as having a duration starting at the sample time and extending to the next sample time.  This conveniently makes associating + combining samples into larger segments a bit easier, and other operations, but is a convention.  I would argue that if a sample has duration, the sample time should instead be the middle of the duration, given the common oversampling and decimation used in most recording systems.

**Comment by etmseed** [comment]
Thanks @chad-earthscope for your interest. I totally agree that samples have no duration. I feel like that with my example above, it is obvious how considering that they have would lead to some inconsistency, because their semantics with regards to the signal would vary with the rate...
If samples are instantaneous values, records have definitely a duration and it has clearly be chosen by convention that this duration extends to the hypothetical next sample after the last. Thus I think this all boils down to chosing if the flags in the record header are set on the record (duration) or on the samples (points in time).
In my opinion, analyzing a signal is analyzing the samples, not some time ranges when there is no data to decide anything. The choice to extend the scope of the analyzis to the date in the middle of two samples should be left to the analyzer and not forced by the miniSEED generator. Yet, once again, I do not know enough use cases and users to be too affirmative. It is only my feeling...

**Comment by etmseed** [comment]
Any update ? Who could decide about that design choice ? A doodle to the community of users ?
